### ISSUE TITLE
[Feature Request] Hyper-parameter optimization in ignite

### ISSUE THREAD

**Issue by walkacross** [pytorch/ignite#181]
Hey guys, do you think it's a good idea that ignite supports hyper-parameter optimization/ finetune? 
I am wondering what's the best practice for finetune in ignite?

1 Do hyper-paramter grid search in run function(like in mnist.py example)? 
this is simple and straight by creating different trainers with for-loop, but may be lack of efficiency.

**Comment by vfdev-5** [comment]
> hyper-paramter grid search in run function(like in mnist.py example)

Hi @walkacross I think your approach is correct, ignite's Engine is not aware of what model with what kind of parameters it runs. 

Today, there is no roadmap on this. Maybe after the first release we could think about this direction in a contrib module. Anyway, I think it could be not difficult to mix ignite with libs like `hyperopt` for random search.

**Comment by vfdev-5** [comment]
@walkacross I close the issue as it seems that I answered your question. But feel free to reopen it or just ask here if you have more questions concerning this issue.

**Comment by karandwivedi42** [comment]
Mostly EarlyStopping + A way to define hyperparameter space + **A way to resume the meta-experiment** + Good logging should do the trick.

**Comment by vfdev-5** [comment]
> A way to resume the meta-experiment 

@karandwivedi42 are you talking about something like #62 ?

Concerning other points, please create more detailed issues and as usual PR are welcome :)

**Comment by alykhantejani** [comment]
I agree with @jasonkriss on this. I think we can provide handlers that might help bridge this gap (checkpointing, early stopping etc.) as well as examples where ignite is used with other libs like `hyperopt`

**Comment by karandwivedi42** [comment]
@alykhantejani @jasonkriss From my experience (at labs and startups with limited GPUs), currently there's no workflow around hyperparameter optimization. Usually, hyperparameter optimization is done by running processes in parallel or one after another with different command line args, along with resuming crashed processes and manually killing processes that aren't doing so good. The results are collected from the logs in a jupyter notebook/table and compared manually. I guess you will agree that it is an way to do this.

I think we can aim to make hyperparameter optimization more plug and play than what I wrote above. I guess even if we don't specifically add a module around hyperparameter optimization, it would be great if you can share how you do hyperparameter optimization: best practices, tools etc. and we can write a tutorial for the same. I think tutorial and example codes have a big impact on how people use pytorch (People often copy and build on top of the example codes to try a new idea).

**Comment by vfdev-5** [comment]
@karandwivedi42 if this can help, take a look on [`skorch`](https://github.com/dnouri/skorch) they have an interface with a grid search for hyperparameter optimization. 

> I think we can aim to make hyperparameter optimization more plug and play than what I wrote above. I guess even if we don't specifically add a module around hyperparameter optimization, it would be great if you can share how you do hyperparameter optimization: best practices, tools etc. and we can write a tutorial for the same. 

It would be great if you could analyze what is missing today and propose some concrete features. Concerning the examples, maybe once the first release is done we can work more on various types of tasks/use-cases to cover by ignite. Personnally, I started a repo on ignite examples to cover some classification use-cases etc. Maybe once here we could provide a curated list of repositories that can be considered as a good practice and so the users can start with that.

**Comment by karandwivedi42** [comment]
> It would be great if you could analyze what is missing today and propose some concrete features. 

Sure. I will raise a new issue after first release.

From my side, this issue can be closed :)
